### PR TITLE
Sso

### PR DESCRIPTION
**Pull request by DavidGeismarLtd**
SSO branch to review.
To test 
- launch ngrok http -region eu -hostname=zedental.eu.ngrok.io 3000
- go on [360 learning](https://zedental.360learning.com/)
- click connection

Still need to add a test

### CODE REVIEW

File: app/services/single_sign_on.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    def initialize(resource)
+      self.resource = resource
+    end
+
+    def ssoable?
+      ssoable_user?
+    end
+
+    def ssoable_user?
```
Comment (ManyBaby) [inline app/services/single_sign_on.rb:29]: By convention predicate methods (ending with a question mark) should alway return a boolean

File: spec/factories/user.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+FactoryBot.define do
```
Comment (ManyBaby) [inline spec/factories/user.rb:1]: Well structured factory

File: db/seeds.rb [hunk 3]
```diff
@@ -36,6 +36,12 @@
                 password: 'password',
                 password_confirmation: 'password',
                 admin: false
+              },
```
Comment (ManyBaby) [inline db/seeds.rb:39]: JP is in da building !

File: config/routes.rb [hunk 4]
```diff
@@ -13,7 +13,8 @@
     resources :registrations
     root to: "users#index"
   end
-  devise_for :users do
+
+  devise_for :users, controllers: {sessions: "sessions"} do
```
Comment (ManyBaby) [inline config/routes.rb:17]: Overriding devise controllers

File: app/services/single_sign_on.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    end
+
+    def url
+      ssoable?
+      "#{ENV['LMS_BASE_URL']}?company=#{ENV['LEARNING_COMPANY_ID']}&jwt=#{token}"
+    end
+
+    private
+
+    attr_accessor :resource
```
Comment (ManyBaby) [inline app/services/single_sign_on.rb:19]: I didn't new that you could have private `attr_accessor` what are the downsides of doing this ?

File: app/services/single_sign_on.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  class UnauthorizedRessource < RuntimeError; end
+  # generates the sso redirect url
+
+  class RedirectService
+    def self.redirect_url(resource)
+      new(resource).url
+    end
+
+    def url
+      ssoable?
```
Comment (ManyBaby) [inline app/services/single_sign_on.rb:13]: I don't think that the concern of SSO is to check for registrations or role of a user.
I would remove that entirely and rely on groups for that as for now the only way to create an account is either to buy a Program or to be an admin.
```suggestion
```

**Review by ManyBaby — APPROVED** [review]
Well played but I think that this won't be the main flow.
We need a way to redirect a authenticated user from zendental.com to 360 without having to go to 360 in a first place.

File: app/services/single_sign_on.rb (hunk 1 above)
Comment (DavidGeismarLtd) [reply app/services/single_sign_on.rb:29]: ok

File: spec/factories/user.rb (hunk 2 above)
Comment (DavidGeismarLtd) [reply spec/factories/user.rb:1]: thanks

File: db/seeds.rb (hunk 3 above)
Comment (DavidGeismarLtd) [reply db/seeds.rb:39]: ya

File: config/routes.rb (hunk 4 above)
Comment (DavidGeismarLtd) [reply config/routes.rb:17]: yes I dont think it's a problem though

Comment (ManyBaby) [reply config/routes.rb:17]: yeah but I'm not familiar with messing around with security

**Pull request merged**